custom/src/main/java/org/kiwix/kiwixmobile/custom/download/CustomDownloadFragment.kt

```diff
@@ -56,24 +57,38 @@ class CustomDownloadActivity : BaseActivity() {
   @Inject lateinit var viewModelFactory: ViewModelProvider.Factory
 
   private val compositeDisposable = CompositeDisposable()
+  override fun inject(baseActivity: BaseActivity) {
+    baseActivity.customActivityComponent.inject(this)
+  }
 
-  override fun onCreate(savedInstanceState: Bundle?) {
+  override fun onCreateView(
+    inflater: LayoutInflater,
+    container: ViewGroup?,
+    savedInstanceState: Bundle?
+  ): View? {
     super.onCreate(savedInstanceState)
-    setContentView(R.layout.activity_custom_download)
-    downloadViewModel.state.observe(this, Observer(::render))
+    val root = inflater.inflate(R.layout.fragment_custom_download, container, false)
+    val activity = requireActivity() as CoreMainActivity
+    downloadViewModel.state.observe(viewLifecycleOwner, Observer(::render))
     compositeDisposable.add(
       downloadViewModel.effects.subscribe(
-        { it.invokeWith(this) },
+        { it.invokeWith(activity) },
         Throwable::printStackTrace
       )
     )
+    return root
+  }
+
+  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+    super.onViewCreated(view, savedInstanceState)
     cd_download_button.setOnClickListener { downloadViewModel.actions.offer(ClickedDownload) }
     cd_retry_button.setOnClickListener { downloadViewModel.actions.offer(ClickedRetry) }
   }
 
   override fun onDestroy() {
     super.onDestroy()
     compositeDisposable.clear()
+    activity?.finish()
```

Comment:
this will kill the activity, we don't want that to happen in onDestroy. Otherwise a completion of download would kill the activity I think

Reply:
I tried this on a physical device, but didn't find anything suspicious or odd

Reply:
Yeah I tested it there and it seems fine. Ah of course

```
class FinishAndStartMain @Inject constructor() : SideEffect<Unit> {
  override fun invokeWith(activity: AppCompatActivity) {
    activity.finish()
    activity.start<CustomMainActivity>()
  }
}
```

That effect triggers when a download completes

Reply:
Thanks for providing the code which makes this possible under the hood, i considered it(launch of mainActivity) as a blackbox before.

Reply:
This a part of our MVI architecture which this screen and a few others adhere to.
The basic idea is that a View sends Actions(ie a user pressed a button) to a ViewModel and receives State(ie this list now has items) from a ViewModel to render.
A ViewModel receives Actions and based on logic in conjunction with the current State triggers a State update or sends off a Side Effect(ie show a toast, navigate somewhere).

It is quite complex at first but an immensely satisfying pattern. It is only going to grow more prevalent on android with compose.

Reply:
Hmm, i can portray it. But why do we have state update named as _SideEffect_, it seems diverging from the root idea, shouldn't it be like _StateEvent_ meaning: the event which is fired off when a change in state is found/observed/received.
Or did i get anything wrong here?

Reply:
State (or at least our States) are the information necessary for a screen to display, the typical thing you might see in tutorials is LCE(Loading Content Error) eg:

A user launches the screen, the intial state is loading so I show a spinner.
The viewmodel performs a web request.
If the request was successful we return a Content state with the listitems, the screen renders this list.
If the request failed then we return an Error state, in our ui we render a button to allow a user to retry
If the user presses that button then we send the action `UserPressedRetry` to do the request

Side effects are transient state, State is a definitive way a screen can display

Reply:
Got it!!

Reply:
Got it!!